In [1]:
import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from gensim.models import Word2Vec
from tabulate import tabulate
from collections import Counter

In [2]:
df = pd.read_csv('news_dataset.csv')

# Extract the 'text' column as a list
dataset = df['text'].dropna().tolist()

print(dataset[:5])

['I was wondering if anyone out there could enlighten me on this car I saw\nthe other day. It was a 2-door sports car, looked to be from the late 60s/\nearly 70s. It was called a Bricklin. The doors were really small. In addition,\nthe front bumper was separate from the rest of the body. This is \nall I know. If anyone can tellme a model name, engine specs, years\nof production, where this car is made, history, or whatever info you\nhave on this funky looking car, please e-mail.', 'I recently posted an article asking what kind of rates single, male\ndrivers under 25 yrs old were paying on performance cars. Here\'s a summary of\nthe replies I received.\n \n \n \n \n-------------------------------------------------------------------------------\n \nI\'m not under 25 anymore (but is 27 close enough).\n \n1992 Dodge Stealth RT/Twin Turbo (300hp model).\nNo tickets, no accidents, own a house, have taken defensive driving 1,\nairbag, abs, security alarm, single.\n \n$1500/year  $500 decut. S

In [15]:
def convert_to_lowercase(text):
    return text.lower()
dataset = df['text'].dropna().tolist()

# Convert each text to lowercase
dataset = [convert_to_lowercase(text) for text in dataset]

# Print only the first 20 items to avoid overloading the console
for i, text in enumerate(dataset[:20], start=1):
    print(f"{i}: {text}")

1: i was wondering if anyone out there could enlighten me on this car i saw
the other day. it was a 2-door sports car, looked to be from the late 60s/
early 70s. it was called a bricklin. the doors were really small. in addition,
the front bumper was separate from the rest of the body. this is 
all i know. if anyone can tellme a model name, engine specs, years
of production, where this car is made, history, or whatever info you
have on this funky looking car, please e-mail.
2: i recently posted an article asking what kind of rates single, male
drivers under 25 yrs old were paying on performance cars. here's a summary of
the replies i received.
 
 
 
 
-------------------------------------------------------------------------------
 
i'm not under 25 anymore (but is 27 close enough).
 
1992 dodge stealth rt/twin turbo (300hp model).
no tickets, no accidents, own a house, have taken defensive driving 1,
airbag, abs, security alarm, single.
 
$1500/year  $500 decut. state farm insurance (t

In [21]:
import re
import string

def remove_punctuation(text):
    if not isinstance(text, str):
        return text
    text = re.sub(r'[\n\r]+', ' ', text)
    punctuation_to_remove = string.punctuation.replace("'", "")
    text = re.sub(r"[{}]".format(re.escape(punctuation_to_remove)), " ", text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text

dataset = [remove_punctuation(text) for text in dataset]

for i, text in enumerate(dataset[:20], start=1):
    print(f"{i}: {text}")

1: i was wondering if anyone out there could enlighten me on this car i saw the other day it was a 2 door sports car looked to be from the late 60s early 70s it was called a bricklin the doors were really small in addition the front bumper was separate from the rest of the body this is all i know if anyone can tellme a model name engine specs years of production where this car is made history or whatever info you have on this funky looking car please e mail
2: i recently posted an article asking what kind of rates single male drivers under 25 yrs old were paying on performance cars here's a summary of the replies i received i'm not under 25 anymore but is 27 close enough 1992 dodge stealth rt twin turbo 300hp model no tickets no accidents own a house have taken defensive driving 1 airbag abs security alarm single 1500 year 500 decut state farm insurance this includes the additional 100 for the 1 000 000 umbrella policy over my car and house the base policy is the standard 100 000 100 0

In [23]:
from nltk.corpus import stopwords

def remove_stopwords(text):
    if isinstance(text, str):
        return ' '.join(word for word in text.split() if word.lower() not in stop_words)
    return text
stop_words = set(stopwords.words('english'))
# Apply all transformations
dataset = [remove_stopwords(text) for text in dataset]

# Print the cleaned dataset
for i, text in enumerate(dataset[:20], start=1):
    print(f"{i}: {text}")

1: wondering anyone could enlighten car saw day 2 door sports car looked late 60s early 70s called bricklin doors really small addition front bumper separate rest body know anyone tellme model name engine specs years production car made history whatever info funky looking car please e mail
2: recently posted article asking kind rates single male drivers 25 yrs old paying performance cars here's summary replies received 25 anymore 27 close enough 1992 dodge stealth rt twin turbo 300hp model tickets accidents house taken defensive driving 1 airbag abs security alarm single 1500 year 500 decut state farm insurance includes additional 100 1 000 000 umbrella policy car house base policy standard 100 000 100 000 300 000 policy required de 2nd defensive driving course 5 less bought car september 1992 company never accident ticket 11 years quoted 2 500 hope helps steve flynn university delaware 45 kevin hope remembered name correctly asked insurance performance cars well last year similar situ

In [25]:
def remove_suffixes(text):
    if isinstance(text, str):
        words = text.split()
        stripped_words = []
        for word in words:
            # Basic handling for common suffixes
            if word.endswith("ing") and len(word) > 4:
                stripped_words.append(word[:-3])
            elif word.endswith("es") and len(word) > 3:
                stripped_words.append(word[:-2])
            else:
                stripped_words.append(word)
        return ' '.join(stripped_words)
    return text

# Apply the new step AFTER the previous processing
dataset = [remove_suffixes(text) for text in dataset]

# Print the final output
for i, text in enumerate(dataset[:20], start=1):
    print(f"{i}: {text}")

1: wonder anyone could enlighten car saw day 2 door sports car looked late 60s early 70s called bricklin doors really small addition front bumper separate rest body know anyone tellme model name engine specs years production car made history whatever info funky look car please e mail
2: recently posted article ask kind rat single male drivers 25 yrs old pay performance cars here's summary repli received 25 anymore 27 close enough 1992 dodge stealth rt twin turbo 300hp model tickets accidents house taken defensive driv 1 airbag abs security alarm single 1500 year 500 decut state farm insurance includ additional 100 1 000 000 umbrella policy car house base policy standard 100 000 100 000 300 000 policy required de 2nd defensive driv course 5 less bought car september 1992 company never accident ticket 11 years quoted 2 500 hope helps steve flynn university delaware 45 kevin hope remembered name correctly asked insurance performance cars well last year similar situation bought car made in

In [37]:
import re
dataset = [text for text in dataset if not re.search(r'\d', text)]

for i, text in enumerate(dataset[:20], start=1):
    print(f"{i}: {text}")

1: ford automobile need information whether ford partially responsible car accidents depletion ozone layer also additional information greatly appreciated thanks sssssoooooooooooo fault thank god louis chevrolet innocent guy diesel otto feel guilty
2: yo watch attributions say appropriate forum discussions whether shoot someone property damage vandalism theft every responsible gun owner realiz limits punishment must fit crime mean think really harmless prank worth kill said situation described punks sett alarms taunt people come could turn ugly quickly worth prepared life potentially line jam
3: agree six hour long stretch behind wheel really make thirsty especially someth caffeine consider fail car cup holder anywhere put cup holder
4: sorry last post server neglected send message
5: appeared today
6: autoweek article car within past six weeks issue diablo vt awd cover naturally remember date issue offhand check anyone interested aamir qazi
7: worry leaks worry way turn damn th take g

In [43]:
!pip install symspellpy
from symspellpy import SymSpell, Verbosity

# Create and load SymSpell
sym_spell = SymSpell(max_dictionary_edit_distance=2, prefix_length=7)
sym_spell.load_dictionary("frequency_dictionary_en_82_765.txt", term_index=0, count_index=1)

def autocorrect_text(text):
    if not isinstance(text, str):
        return text
    corrected_words = []
    for word in text.split():
        suggestions = sym_spell.lookup(word, Verbosity.CLOSEST, max_edit_distance=2)
        corrected_word = suggestions[0].term if suggestions else word
        corrected_words.append(corrected_word)
    return ' '.join(corrected_words)

# Apply correction to all dataset items
corrected_dataset = [autocorrect_text(text) for text in dataset]

# Show first 20 results
for i, text in enumerate(corrected_dataset[:20], start=1):
    print(f"{i}: {text}")


2025-05-15 18:09:09,559: E symspellpy.symspellpy] Dictionary file not found at frequency_dictionary_en_82_765.txt.


   ---------------------------------------- 0.0/2.6 MB ? eta -:--:--
   ---------------- ----------------------- 1.0/2.6 MB 5.6 MB/s eta 0:00:01
   -------------------------------- ------- 2.1/2.6 MB 5.3 MB/s eta 0:00:01
   ---------------------------------------- 2.6/2.6 MB 5.2 MB/s eta 0:00:00
1: ford automobile need information whether ford partially responsible car accidents depletion ozone layer also additional information greatly appreciated thanks sssssoooooooooooo fault thank god louis chevrolet innocent guy diesel otto feel guilty
2: yo watch attributions say appropriate forum discussions whether shoot someone property damage vandalism theft every responsible gun owner realiz limits punishment must fit crime mean think really harmless prank worth kill said situation described punks sett alarms taunt people come could turn ugly quickly worth prepared life potentially line jam
3: agree six hour long stretch behind wheel really make thirsty especially someth caffeine consider fai

In [45]:
tokenized_dataset = [doc.split() for doc in dataset]
word2vec_model = Word2Vec(sentences=tokenized_dataset, vector_size=100,
window=5, min_count=1, workers=4)

In [49]:
import numpy as np

embedding_dim = word2vec_model.vector_size  # e.g., 300

X = []
for doc in dataset:
    vectors = [word2vec_model.wv[word] for word in doc.split() if word in word2vec_model.wv]
    if vectors:
        doc_vector = np.mean(vectors, axis=0)
    else:
        doc_vector = np.zeros(embedding_dim)  # or handle differently (skip, nan, etc.)
    X.append(doc_vector)

X = np.array(X)

In [55]:
k = 2  # number of clusters
km = KMeans(n_clusters=k, random_state=42)
km.fit(X)

# Predict cluster for each document
y_pred = km.predict(X)

# Prepare and print only first 20 results
table_data = [["Document", "Predicted Cluster"]]
table_data.extend([[doc, cluster] for doc, cluster in zip(dataset[:20], y_pred[:20])])

print(tabulate(table_data, headers="firstrow"))

Document                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Predicted Cluster
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  -------------------


In [57]:
# Calculate purity
total_samples = len(y_pred)
cluster_label_counts = [Counter(y_pred)]
purity = sum(max(cluster.values()) for cluster in cluster_label_counts) / total_samples
print("Purity:", purity)


Purity: 0.6428257042253521


In [65]:
from gensim.models.coherencemodel import CoherenceModel

k = 2
clusters = y_pred

# Create a list of tokenized documents (list of words)
tokenized_docs = [doc.split() for doc in dataset]

# For each cluster, aggregate words
cluster_words = []
for cluster_id in range(k):
    # Combine all words from docs in this cluster
    words = []
    for i, c in enumerate(clusters):
        if c == cluster_id:
            words.extend(tokenized_docs[i])
    # Get top N words
    top_words = [word for word, freq in Counter(words).most_common(20)]
    cluster_words.append(top_words)

# Now compute coherence score for each cluster's top words
# We need a gensim Dictionary and corpus for the whole dataset
from gensim.corpora.dictionary import Dictionary

dictionary = Dictionary(tokenized_docs)
corpus = [dictionary.doc2bow(doc) for doc in tokenized_docs]

coherence_scores = []
for top_words in cluster_words:
    cm = CoherenceModel(
        topics=[top_words], 
        texts=tokenized_docs, 
        dictionary=dictionary, 
        corpus=corpus, 
        coherence='c_v'
    )
    score = cm.get_coherence()
    coherence_scores.append(score)

for idx, top_words in enumerate(cluster_words):
    cm = CoherenceModel(
        topics=[top_words],
        texts=tokenized_docs,
        dictionary=dictionary,
        corpus=corpus,
        coherence='c_v'
    )
    score = cm.get_coherence()
    print(f"Cluster {idx} top words: {top_words}")
    print(f"Cluster {idx} coherence score: {score:.4f}\n")

Cluster 0 top words: ['would', 'one', 'like', 'us', 'people', 'go', 'also', 'think', 'know', 'get', 'use', 'could', 'time', 'right', 'well', 'even', 'see', 'new', 'may', 'good']
Cluster 0 coherence score: 0.3761

Cluster 1 top words: ['would', 'one', 'people', 'think', 'know', 'like', 'get', 'go', 'us', 'could', 'say', 'god', 'good', 'time', 'even', 'well', 'see', 'also', 'way', 'make']
Cluster 1 coherence score: 0.4039



Cluster 1 is slightly more coherent (0.404) than Cluster 0 (0.376), but both scores are low, likely due to common words dominating. Better preprocessing is needed for clearer topics.

Muhammad Irfan Ammar bin Zaharudin(SW01082532)
Wan Emir Hussein bin Wan Rahim (SN01082115)